### ISSUE TITLE
CLI: SourceMap sources path should be relative

### ISSUE THREAD

**Issue by am11** [jashkenas/coffeescript#3296]
Relative to the location of compiled map and `.coffee` or absolute URL on disk, when compiled with Node.

Consider the following structure:

``` bash
website_root
        \
        assets
        \
        coffee
               \
                work.coffee
        \
        javascript
               \
                 work.js
               \
                 work.js.map
```

It was result of this command:

``` c#
> node "node_modules\iced-coffee-script\bin\coffee"
  --bare --runtime inline --map
  --output "C:\projects\test\website_root\javascript\" 
  --compile  "C:\projects\test\website_root\coffee\work.coffee"
```

Now, the `sources[]` inside the `.map` file looks like:

``` json
"sources": [
  "worker.coffee"
],
```

While expected result is:

**relative path**

``` json
"sources": [
  "../coffee/worker.coffee"
],
```

**or absolute path**

``` json
"sources": [
  "c:\projects\test\website_root\coffee\worker.coffee"
],
```

**or optional switch for either:**

Note: LESS provides `--source-map-basepath={relative or absolute path}` argument and you can supply desired path to .map file as well. That might worth checking out in this situation.

---

<h3>Proposed Solution:</h3>

LESS provides six compiler arguments to deal with source map in general, out of which four effects on file paths, viz;

``` bash
  --source-map[=FILENAME]      # Outputs a v3 sourcemap to the filename (or output filename.map)
  --source-map-rootpath=X         # adds this path onto the sourcemap filename and less file paths
  --source-map-basepath=X       # Sets sourcemap base path, defaults to current working directory.
  --source-map-url=URL             # the complete url and filename put in the less file
  --source-map-less-inline          # puts the less files into the map instead of referencing them
  --source-map-map-inline         # puts the map (and any less files) into the output css file
```

_(first four)_

I personally find it very confusing and one may not get it unless experiment with it for a while.

Having said that, considering our aim is to keep the _end result_ consistent, that is; **the related .coffee, .map and .js can be saved at different locations**, we can fix it by providing "value" in existing compiler switch; `--map[=FileNameWithPath]`. After that, calculate the following relative paths:
- from `.css` to `.map`; for sourceMap comment in CSS file.
- from `.map` to `.css`; for the value of "file" [member / property*](http://json-schema.org/latest/json-schema-core.html#anchor4) in `.map` JSON.
- from `.map` to each `.less` file; for sources array in `.map` JSON.

[This is how](https://github.com/madskristensen/WebEssentials2013/blob/master/EditorExtensions/Helpers/FileHelpers.cs#L171-L214) we calculated relative paths in .NET independent of FileSystem API. We can write a similar method in JavaScript.

For Windows, there is C++ file system function to resolve relative paths; [`PathRelativePathTo()`](http://msdn.microsoft.com/en-us/library/windows/desktop/bb773740.aspx) in `shlwapi.dll`. I am sure for other OS and languages, there must be a similar low-level implementation to calculate the shortest path between two files on disc. 

For further references please check the LESS implementation and see this discussion for various approaches to solve relative URL issues: https://github.com/madskristensen/WebEssentials2013/issues/134.

**Comment by joliss** [comment]
FYI, we're trying to get source maps to "just work" for the CoffeeScript plugin for the Broccoli build tool (https://github.com/joliss/broccoli-coffee/pull/9), and this bug is blocking us. Would love to see this fixed!

**Comment by jashkenas** [comment]
Let's fix it!

**Comment by am11** [comment]
Yay!

**Comment by Hubro** [comment]
I just had to do some pretty awkward stuff in my coffee build script (involving copying my whole source directory into my static js folder and changing working directory) to get the source mapping feature to work. Being able to override the root path and base path would have saved me a couple of headaches. 

Has there been any progress in that regard?

**Comment by GeoffreyBooth** [comment]
See also #3075.

**Comment by GeoffreyBooth** [comment]
I just tested this, and I’m unable to reproduce (which isn’t all that surprising, as it’s four years later):

```bash
➜ coffee -v
CoffeeScript version 2.0.0-beta4
➜ mkdir -p website_root/assets website_root/coffee website_root/js
➜ echo 'console.log Date.now()' > ./website_root/coffee/work.coffee
➜ coffee --bare --map --output ./website_root/js --compile ./website_root/coffee/work.coffee
➜ cat ./website_root/js/work.js.map
{
  "version": 3,
  "file": "work.js",
  "sourceRoot": "../..",
  "sources": [
    "website_root/coffee/work.coffee"
  ],
  "names": [],
  "mappings": ";AAAA,OAAO,CAAC,GAAR,CAAY,IAAI,CAAC,GAAL,CAAA,CAAZ"
}
```

Note the relative path in the `sources` array. It could also be that something in Node’s `fs` module changed to fix this; I’m running Node 8.4.0. I’m also on OS X, in case perhaps this was a Windows issue (or perhaps a Windows issue in an older version of Node).

If this _is_ still an issue and someone can write some steps to reproduce, please post those steps in a comment and I’ll reopen.

At this point we don’t plan to add any more command-line flags, so there won’t be an option to get `sources` output as absolute paths. Creating a build script to convert these relative paths to absolute ones, though, shouldn’t be difficult. Here’s [an example from Coffeeify rewriting the `sources` field](https://github.com/jnordberg/coffeeify/blob/f2c99868d76e4ccf9cd1ef856d4c72f9a9026642/index.js#L67).
